File: elixir/day2/lib/part2.ex
defmodule Part2 do
  def parse_line(line) do
    line
    |> String.split(";")
    |> Enum.map(fn game ->
      game
      |> String.split(",")
      |> Enum.reduce(%{}, fn bag, acc ->
        [count, colour] = String.split(String.trim(bag))
        Map.put(acc, String.to_atom(colour), String.to_integer(count))
      end)
    end)
  end

  def parse_game(game) do
    {red, green, blue} =
      Enum.reduce(game, {0, 0, 0}, fn turn, {max_red, max_blue, max_green} ->
        reds = turn[:red] || 0
        blues = turn[:blue] || 0
        greens = turn[:green] || 0

        {max(reds, max_red), max(blues, max_blue), max(greens, max_green)}
      end)

    red * green * blue
  end

  def part2(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.reduce(0, fn line, acc ->
      [_, _, game] = Regex.run(~r/Game (\d+): (.*)/, line)
      turns = parse_line(game)
      acc + parse_game(turns)
    end)
  end
end

File.read!("input.txt")
|> Part2.part2()
|> IO.inspect(label: "part2")

File: ruby/day2/part2.rb
# frozen_string_literal: true

def parse_line(line)
  line.split(';').map do |game|
    game.split(',').to_h do |bag|
      count, colour = bag.strip.split
      [colour.to_sym, count.to_i]
    end
  end
end

def parse_game(game)
  max_red = 0
  max_blue = 0
  max_green = 0

  game.each do |turn|
    max_red = [turn[:red], max_red].max unless turn[:red].nil?
    max_blue = [turn[:blue], max_blue].max unless turn[:blue].nil?
    max_green = [turn[:green], max_green].max unless turn[:green].nil?
  end

  max_red * max_blue * max_green
end

def part2(input)
  input.split("\n").reduce(0) do |result, line|
    game = line.match(/Game \d+: (.*)/).captures.first
    turns = parse_line(game)
    result + parse_game(turns)
  end
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  p part2(input)
end
